resources/type-system/strict-inference.md

```diff
@@ -297,12 +297,14 @@ void main() {
 
 ### Function return types
 
-Declaring a recursive local function, a top-level function, a method, a
-typedef, a generic function type, or a function-typed function parameter
-without a return type is an inference failure. The return type of non-recursive
-local functions can always be inferred from downwards or upwards inference, as
-it will have a body, and the return type of the body is known (even if there
-are inference failures within).
+Declaring a recursive local function, a top-level function, a static method, a
+non-overriding instance method, a typedef, a generic function type, or a
```

Comment:
"a typedef" is traditionally called a "a type alias".

The "a generic function type" case is different from the rest. Every other thing you mention is the syntactic category of a declaration, which makes sense since the sentence starts with "Declaring a ...". A "generic function type" is not a declaration, so it's not as clear what it means to "declare" one.

Why is it important that the function type is generic?

Reply:
Hmm, I may just be abusing the analyzer's AST for this concept, [GenericFunctionType](https://pub.dev/documentation/analyzer/0.37.0/dart_ast_ast/GenericFunctionType-class.html). It's not important that it be declared with type arguments. E.g.

```dart
void f6(callback()) {                 // Inference failure
  callback();
}
```

The issue there is that the return type of `callback` is implicitly `dynamic`; no inference is carried out. Is this just called a "function type?"